### ISSUE TITLE
Command-line Arguments for /Applications/Docker.app/Contents/MacOS/Docker 

### ISSUE THREAD

**Issue by rfay** [docker/for-mac#2359]
### Expected behavior

Docker for Mac 18.09.* [edit 2019-02-03]

Could you please let me know the allowed command-line arguments for /Applications/Docker.app/Contents/MacOS/Docker? I know that these sort-of work:
```
/Applications/Docker.app/Contents/MacOS/Docker --unattended
/Applications/Docker.app/Contents/MacOS/Docker --quit-after-install --unattended
```
but only because I happened to find a reference to them in https://github.com/docker/for-mac/issues/882#issuecomment-354661036

Currently I can sudo run /Applications/Docker.app/Contents/MacOS/Docker --unattended
 in a headless environment, but only with sudo, and really want to know how to do it to set up docker for a normal user. (For testing on CircleCI)

### Information

  - Full output of the diagnostics from "Diagnose & Feedback" in the menu
  - A reproducible case if this is a bug, Dockerfiles FTW
  - Page URL if this is a docs issue or the name of a man page

### Steps to reproduce the behavior

  1. ...
  2. ...

**Comment by GabLeRoux** [comment]
I just learned about the `unattended` a bit like you and was curious if I could run Docker for mac on Travis, but that's not possible:

```
+brew cask install docker
==> Satisfying dependencies
==> Downloading https://download.docker.com/mac/stable/23751/Docker.dmg
==> Verifying checksum for Cask docker
==> Installing Cask docker
==> Moving App 'Docker.app' to '/Applications/Docker.app'.
🍺  docker was successfully installed!
+docker system info
+sudo /Applications/Docker.app/Contents/MacOS/Docker --unattended
+sleep 1
Settings: >>> New settings saved <<<
Settings: linuxDaemonConfigCreationDate : 2018-07-26 15:40:23 +0000
Settings: <<< New settings saved >>>
CrashReport: stderr redirected to /var/root/Library/Group Containers/group.com.docker/DockerAppStderr.txt
FeaturesStateLoader: featuresState saved
AppDelegate: Application version: 18.03.0-ce-mac60 (23751)
AppDelegate: Channel: stable
AppDelegate: Administrator user: true
OSX Version: Version 10.13.3 (Build 17D47)
Terminator: Incompatible CPU detected.
We are sorry, but your hardware is incompatible with Docker for Mac.
Docker requires a processor with virtualization capabilities and hypervisor support.
To learn more about this issue see:
[https://docs.docker.com/docker-for-mac/troubleshoot](https://docs.docker.com/docker-for-mac/troubleshoot/#incompatible-cpu-detected)
```

**Comment by rfay** [comment]
For the basic question of "How to install docker-for-mac from the command line" this works:

https://github.com/drud/ddev/blob/fe1b8ddbfeb322e4966b31c5654362f6a0b022be/.circleci/macos_circle_vm_setup.sh#L11-L13

**Comment by rfay** [comment]
Unfortunately, this ceases working in Docker Desktop for Mac 2.1.0.0 (docker v19.03.1). Now you get the error message:
```
DockerApplication: Docker.app starting for root
Terminator: Running Docker Desktop as root is dangerous. Please run it as a regular user.
```

Which breaks our CI :( For now, I can download earlier version, but Docker folk... please tell us how to do an unattended installation!

**Comment by apolishch** [comment]
Any plans to get this addressed? I'm still being hit with this in CI

**Comment by msmoljan** [comment]
Fixing this would be extremely helpful for our CI workflow too. 

Currently I'm using this workaround from @rfay as a temporary fix:
https://github.com/drud/ddev/pull/1748/files#diff-19288f650af2dabdf1dcc5b354d1f245

**Comment by rfay** [comment]
Thanks for fixing that @cweagans -  Would love to hear from the docker team on this simple issue, thanks! https://twitter.com/randyfay/status/1230757005792432128

**Comment by FranDepascuali** [comment]
I wrote about setting up docker headless in this post, it may be of help:
https://deep-thought.netlify.com/posts/using-docker-fastlane

**Comment by telamonian** [comment]
Here's a way to get a working copy of the latest Docker Desktop app for Mac installed from the command line:

```bash
#!/usr/bin/env bash

# refs:
# https://github.com/MicrosoftDocs/vsts-docs/issues/3784
# https://forums.docker.com/t/docker-for-mac-unattended-installation/27112

brew cask install docker
# allow the app to run without confirmation
xattr -d -r com.apple.quarantine /Applications/Docker.app

# preemptively do docker.app's setup to avoid any gui prompts
sudo /bin/cp /Applications/Docker.app/Contents/Library/LaunchServices/com.docker.vmnetd /Library/PrivilegedHelperTools
sudo /bin/cp /Applications/Docker.app/Contents/Resources/com.docker.vmnetd.plist /Library/LaunchDaemons/
sudo /bin/chmod 544 /Library/PrivilegedHelperTools/com.docker.vmnetd
sudo /bin/chmod 644 /Library/LaunchDaemons/com.docker.vmnetd.plist
sudo /bin/launchctl load /Library/LaunchDaemons/com.docker.vmnetd.plist
```

and then you can run the app and wait for it to install/set up with this script:

```bash
#!/usr/bin/env bash

# refs:
# https://stackoverflow.com/a/35979292/425458

[[ $(uname) == 'Darwin' ]] || { echo "This function only runs on macOS." >&2; exit 2; }

echo "-- Starting Docker.app, if necessary..."

/Applications/Docker.app/Contents/MacOS/Docker &

# Wait for the server to start up, if applicable.
i=0
while ! docker system info &>/dev/null; do
  (( i++ == 0 )) && printf %s '-- Waiting for Docker to finish starting up...' || printf '.'
  sleep 1
done
(( i )) && printf '\n'

echo "-- Docker is ready."
```

This is what I needed to do in order to get Docker running during my Mac CI tests (on Azure Pipelines).

My install hack above is completely ridiculous, and there should be a better way. The old way with the cute command line arguments:

```
/Applications/Docker.app/Contents/MacOS/Docker --quit-after-install --unattended
```

was so, so much better. Can we please get this back?

I think a lot of people need this in order to get their CI test suites to pass. The use case is for when you need to run a docker that exposes a service (such as a local S3 server) to the test code.

**Comment by telamonian** [comment]
I think a lot of people need this in order to get their CI test suites to pass. The use case is for when you need to run a docker that exposes a service (such as a local S3 server) to the test code.

@satrapu seems to have also gone through similar issues with docker+mac+azure. Some useful info:

https://crossprogramming.com/2019/12/27/use-docker-when-running-integration-tests-with-azure-pipelines.html#self-managed-docker-containers
https://github.com/satrapu/aspnet-core-logging/blob/8176a9569da56934f83e01b37648d58300198d1e/Build/start-docker-on-macOS.sh

**Comment by joemiller** [comment]
I've had success with @telamonian's script above on github actions. I had to add `--no-quarantine` to the `brew cask install docker` to get it to work after my initial attempt ran into an error:

```
ditto: WARNING: bom /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/20200524-2091-13atkh4.bom contains no files, nothing copied
```

**Comment by rfay** [comment]
https://github.com/docker/for-mac/issues/882 just got closed by docker-desktop-robot, but it's mostly a duplicate of this one.

**Comment by myitcv** [comment]
@telamonian thank you for your comment in https://github.com/docker/for-mac/issues/2359#issuecomment-607154849. Works perfectly but for one thing. I can't see how to configure the resources for the Docker Desktop in an automated/headless way. By default Docker Desktop uses half the number of available CPU cores, and a fraction of the memory. I would like to remove those constraints, giving Docker access to all available resources (my setup is CI via GitHub Actions). Does anyone know how to do this?

**Comment by satrapu** [comment]
Hey @telamonian, thanks for the solution described inside [your comment](https://github.com/docker/for-mac/issues/2359#issuecomment-607154849)!
Since February 1st my solution no longer works due to reasons beyond my understanding, so it's great yours does.
You just made my day today!

**Comment by myitcv** [comment]
Hi @telamonian, thanks again for this wonderful solution. However, as of exactly 4 days ago, this solution has stopped working. 

We use the approach you describe in https://github.com/play-with-go/play-with-go/blob/2c980fc5b1956bb05f259b986dd18d9f58efe869/_scripts/macCISetup.sh. But 4 days ago step that waits for the server to startup:

https://github.com/play-with-go/play-with-go/blob/2c980fc5b1956bb05f259b986dd18d9f58efe869/_scripts/macCISetup.sh#L45-L51

never completes.

Has anyone else seen something like this?

**Comment by MaceWindu** [comment]
It @myitcv  should be caused by https://github.com/docker/for-mac/issues/5406. E.g. we still get 3.2.0 from homebrew for some reason even if it already have 3.2.1 (https://github.com/Homebrew/homebrew-cask/blob/master/Casks/docker.rb).
You could try to specify version explicitly

**Comment by myitcv** [comment]
Thanks very much, @MaceWindu - we needed a `brew update` step in our script as it turns out (which makes sense with hindsight)

**Comment by telamonian** [comment]
Here's how to get a working copy of the very latest Docker Desktop app for Mac installed from the command line:

```bash
#!/usr/bin/env bash

# refs:
# https://github.com/MicrosoftDocs/vsts-docs/issues/3784
# https://forums.docker.com/t/docker-for-mac-unattended-installation/27112

brew install --cask docker
# allow the app to run without confirmation
xattr -d -r com.apple.quarantine /Applications/Docker.app

# preemptively do docker.app's setup to avoid any gui prompts
sudo /bin/cp /Applications/Docker.app/Contents/Library/LaunchServices/com.docker.vmnetd /Library/PrivilegedHelperTools

# the plist we need used to be in /Applications/Docker.app/Contents/Resources, but
# is now dynamically generated. So we dynamically generate our own
sudo tee "/Library/LaunchDaemons/com.docker.vmnetd.plist" > /dev/null <<'EOF'
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>com.docker.vmnetd</string>
	<key>Program</key>
	<string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
	<key>ProgramArguments</key>
	<array>
		<string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
	<key>Sockets</key>
	<dict>
		<key>Listener</key>
		<dict>
			<key>SockPathMode</key>
			<integer>438</integer>
			<key>SockPathName</key>
			<string>/var/run/com.docker.vmnetd.sock</string>
		</dict>
	</dict>
	<key>Version</key>
	<string>59</string>
</dict>
</plist>

EOF

sudo /bin/chmod 544 /Library/PrivilegedHelperTools/com.docker.vmnetd
sudo /bin/chmod 644 /Library/LaunchDaemons/com.docker.vmnetd.plist
sudo /bin/launchctl load /Library/LaunchDaemons/com.docker.vmnetd.plist

sleep 5
```

and then you can run the app and wait for it to install/set up with this script:

```bash
#!/usr/bin/env bash

# refs:
# https://stackoverflow.com/a/35979292/425458

[[ $(uname) == 'Darwin' ]] || { echo "This function only runs on macOS." >&2; exit 2; }

echo "-- Starting Docker.app, if necessary..."

open -g -a Docker.app || exit

# Wait for the server to start up, if applicable.
i=0
while ! docker system info &>/dev/null; do
  (( i++ == 0 )) && printf %s '-- Waiting for Docker to finish starting up...' || printf '.'
  sleep 1
done
(( i )) && printf '\n'

echo "-- Docker is ready."
```

This is a slightly more ridiculous update to my original ridiculous hack: https://github.com/docker/for-mac/issues/2359#issuecomment-607154849

**Comment by satrapu** [comment]
> 
> 
> Here's how to get a working copy of the very latest Docker Desktop app for Mac installed from the command line:
> ...

@telamonian, I confirm your script works.
You, Sir, you just made my day today!

**Comment by myitcv** [comment]
@telamonian thanks for the ping above. Unfortunately the approach you describe in https://github.com/docker/for-mac/issues/2359#issuecomment-853420567 does not appear to be working for me via GitHub actions. What's worse is that I can't debug this on my local mac because it's old (10.12). Is it possible there is still a UI prompt of some sort that is causing my CI workflow to hang?

**Comment by armenzg** [comment]
@myitcv Use `61` instead of `59`. It seems that the dynamically generated `com.docker.vmnetd.plist` changes overtime. (In reference to @telamonian code snippet in [here](https://github.com/docker/for-mac/issues/2359#issuecomment-853420567))

Anyone know where that 61 comes from?
Or anyone know how to `brew install --cask docker` with a specific version? (By doing so, I can stick to a specific Docker version and hopefully the `plist` does not change).

```diff
-	<string>59</string>
+ 	<string>61</string>
```

**Comment by jasoncodes** [comment]
It looks like you can get the vmnetd version number from `Contents/Info.plist`:

```
$ defaults read /Applications/Docker.app/Contents/Info.plist VmnetdVersion
61
```

**Comment by armenzg** [comment]
You would think they have some intention of fixing this.
I think I'm throwing the towel and I'm going to find some other alternative to running containers on Docker.

This is the latest UI prompt and I don't know how to bypass it:
<img width="808" alt="image" src="https://user-images.githubusercontent.com/44410/131721243-ef30c0fc-24ed-4346-9eeb-6ed6f0b721f0.png">

**Comment by MaceWindu** [comment]
oh well, we already disabled our macos pipelines due to unbearable slowlyness. Probably it is time to remove them completely and just forget about them

**Comment by knksmith57** [comment]
@gwynne's work around is awesome and we can confirm it is working for us on macOS `11.16.1` w/ Docker Desktop `4.1.0` 🚀  (huge thanks for that! 🙏 )

**note**: after a system reboot, the license accept didn't "stick" and the re-launched `Docker.app` prompted for license acceptance. I'm guessing if one always opened `Docker.app` with the `--unattended` and `--accept-license` flags that this would continue to Just Work™️, but note that launching with this workaround doesn't appear to permanently dismiss the prompt.

**Comment by joffrey-bion** [comment]
@gwynne's solution did work for me as well, although indeed it took 9m49s to run 😭 
The previous script I was using took about 1m30s, but well at least this new one works.

Every independent developer needs to do this on Github Actions, why not just allow them to preinstall the software?

**Comment by armenzg** [comment]
I filed it [here](https://github.com/actions/runner/issues/1456) in case you want to up-vote it.

**Comment by MaceWindu** [comment]
@armenzg , that's a wrong place to request it as it's blocked on Docker side by their EULA:
https://github.com/actions/virtual-environments/issues/4399#issuecomment-957176340
https://github.com/actions/virtual-environments/issues/2150#issuecomment-735795573

**Comment by armenzg** [comment]
Thanks @MaceWindu !

**Comment by sparty02** [comment]
the [above comment](https://github.com/docker/for-mac/issues/2359#issuecomment-943131345) worked great, but I happened to have `HOMEBREW_CASK_OPTS="--appdir=~/Applications"`, so the app directory where the docker cask gets installed is a bit different.  In that respect, I updated the script a bit to lookup the appdir.

```bash
#!/usr/bin/env bash

brew install --cask docker
docker_app_path=$(brew list --cask docker | grep '==> App' -A1 | tail -n 1 | awk '{ print $1 }')

sudo "$docker_app_path"/Contents/MacOS/Docker --unattended --install-privileged-components
open -a "$docker_app_path" --args --unattended --accept-license
while ! "$docker_app_path"/Contents/Resources/bin/docker info &>/dev/null; do sleep 1; done
```

**Comment by stephen-turner** [comment]
This is mostly fixed in Docker Desktop 4.7.0. Let's move the discussion to docker/roadmap#80.
